### ISSUE TITLE
Bottleneck in freezing/summarizing computed columns?

### ISSUE THREAD

**Issue by jsvine** [saulpw/visidata#100]
Hi, and thanks for all the great work on VisiData!

I've noticed that freezing and summarizing computed columns in VisiData sometimes takes much longer than expected (given the seemingly-analogous operations in plain Python).

For the purposes of this issue, I've tried to come up with a minimalist example of what I've noticed. I'm attaching the three key files here: [visidata-bottleneck-example.zip](https://github.com/saulpw/visidata/files/1603574/visidata-bottleneck-example.zip)

One file is a single-column CSV, where each of the 100,000 rows contains the same date string ("2018-01-03").

I'm also attaching a VisiData command log, which loads the CSV, assigns the column to be of the "date" type, creates a computed column, `=date.year`, and then tries freezing that column.

When I run `vd --batch --play cmdlog-dates-year-freeze.vd`, the process takes about 50s-60s. Ditto if, instead of freezing, I run a frequency table on the computed column.

When I run what *seems to me* (though I could definitely be mistaken) an equivalent Python script, it takes just over one second. I've attached the script, but here's the full text of it, for reference:

```python
import datetime

print("Creating date strings")
date_strings = [ "2018-01-03" for _ in range(100000) ]

print("Converting strings to datetimes")
dts = [ datetime.datetime.strptime(s, "%Y-%m-%d")
    for s in date_strings ]

print("Converting datetimes to years")
years = [ dt.year for dt in dts ]
print(years[:5])
```

Do you know what's going on? Am I doing something silly in VisiData?

**Comment by saulpw** [comment]
There are several things I can think of.  This is a good time to mention the Threads Sheet, which you can go to with `^T`.  This conveniently lists all of the separate threads that have been spawned for long-running tasks, like e.g. freezing a column.  Each row shows the amount of time that the thread took (or is currently taking), and in this case, you're totally right, that the freeze itself is taking most of the time.

You can also press `ENTER` on a thread to open a sheet with a performance profile of that thread in text format.  This is a pretty basic output, with the calls generally in a kind of call-stack order (if there is one primary bottleneck).  For more sophisticated analysis, you can press `z^S` (`z Ctrl-S`) to save the profile as a .prof file, which can be converted with pyprof2calltree and then viewed with kcachegrind (see this [doc on profiling](http://docs-python.readthedocs.io/en/latest/profiling.html)).

So some amount of the time you're seeing is just VisiData overhead (e.g. `getTypedValue`), but most of the time is actually spent in the `python-dateutil` parser.  This is a great library for parsing dates in a wide variety of formats--and it can usually autodetect the format based on the input.  From experience I have developed a strong preference for keeping dependencies to an absolute minimum, but date parsing is so hard and python-dateutil is so useful that I decided to allow it as VisiData's sole external dependency (aside from individual loader requirements).

But that fuzzy autodetecting of the date format is time-consuming, much moreso than just using the builtin `strptime` with a known precise format string, as your script does; and `strptime` doesn't allow any deviations or even trailing characters in the input.  So that was my first guess about why VisiData takes so much longer than your script, although the magnitude of the difference still surprised me.

[This is why freezing a column is sometimes a good idea, as you noticed.  The input is reparsed and each column is recalculated every time any value is needed, so if you're going to use an expensive column multiple times, it's better to calculate it only once.]

It seems that if you have a lot of data and you know the precise input date format, that you should be able to use the much simpler strptime for faster parsing, though.  So I went and added that last night in a private branch: first it tries to strptime with `options.disp_date_fmt` (perhaps this should be a separate option), and if that fails, then it falls back to the more general python-dateutil parser.  To my complete surprise, this did not significantly improve the performance!

Looking at the thread profile after this change, most of the time is still spent in the parser, this time just  with the builtin strptime; and that time in strptime alone is 3x the time that your script takes to run in its entirety!  So something else must also be going on here.

I wondered if it was a poor interaction of multiple threads with the Global Interpreter Lock (GIL), but a quick test in VisiData without threads showed a performance improvement of only 2% or so.

My current theory is that your script is allowing the compiler and/or strptime to optimize aggressively, by using the same input string (and possibly date format string) repeatedly.  Modifying your script to read each date from the `dates.csv` file did have a significant performance impact, but it was still not enough to account for the 3x slowdown that we both are seeing.

At this point, I think there are two things going on here.  First, it is likely that VisiData is not allowing strptime to optimize fully; once that's fixed, then adding the strptime fastpath to the date parser will be a nice performance improvement for those times when you can specify the exact format string to the parser.

Secondly, though, I think there is a performance bug in the replay mechanism, since I've also seen a weird 5x slowdown in e.g. our xlsx test (which just loads a single spreadsheet).  Finding and fixing this bug will be beneficial across the board.

So ultimately I think this slowdown is these things in combination:

1. Your script is over-optimized and not representative of what actually needs to happen;
2. VisiData introduces some some amount of architectural overhead which can't really be avoided, and for maximum performance you will still need to write custom Python code;
3. python-dateutil is very handy but also time-consuming; possibly some cleverness could be used there to cache the result of autodetection;
4. A replay issue is making the specific repro test take a lot more time than it needs to.

I'm going to spend some time later today/tomorrow trying to figure out and hopefully fix the issues that I can.

In the meantime, knowing that the date converter is slow, I think you should be able to get much better performance with a slightly different workflow: if all you want is the year of an ISO-formatted date, you can regex split (with `:`) on `-` and set the type of the first new column to `int`.

**Comment by saulpw** [comment]
There was one thing I missed that was hiding in plain sight:

> press ENTER on a thread to open a sheet with a performance profile

`options.profile_threads` is set to True by default.  Setting it to False (which will now be the default) apparently reduces the time for VisiData async threads (should be not just freezing, but any action that has a progress indicator) by 50%.  Now the rest of the times correlate (except for the replay delay).

I apologize for not thinking to disable profiling by default long before now.  Thanks for pointing this out before 1.0, @jsvine!   

So, here are the results (all times in seconds on my machine):

- 0.1  script using now()
- 0.5  script using strptime
- 3.4  script using dateutil.parser

- 1.0  vd returning now()
- 1.2  vd using strptime
- 4.4  vd with dateutil.parser
- 5.5  vd first trying strptime with invalid format, falling back to dateutil.parser
- 8.4  vd with dateutil.parser and profiling

So, the largest culprit was the thread profiling, and then python-dateutil.  The architectural overhead of VisiData adds half a second, maybe a whole second.  The replay issue is still at-large, but will be treated as a separate issue.

I was wrong about your script though.  It was not over-optimizing, I made a mistake when timing a change I had made to it (I ran it with profiling).

I'm going to keep the current behavior with python-dateutil and leave out the strptime fastpath, unless it is super important to someone.  At some point we could cache (per-column) the internal parser that dateutil.parser discovers.

Thanks again for filing this issue.

**Comment by saulpw** [comment]
Also 1653a9206191edf5a482ec96f80e0ba5b49e4c1d.

**Comment by saulpw** [comment]
The replay issue was `VisiData.sync()` spinning aggressively checking unfinished threads and waiting for them to finish.  Adding a `sleep()` there to yield to threads doing actual work fixes it for both live and batch modes.  In addition, async threads are now disabled for batch mode.  These are commits ffc4dbb8d6dc134a23b9cbfdf3b174d54956b7a9 and f178500b3b2139303a96134ab680d899ff0d293f, which will be cherry-picked to stable for 1.0.

**Comment by jsvine** [comment]
Thanks so much for looking into this @saulpw, and absolutely fascinating to hear what you found. In hindsight, I should have provided a more comparable example, e.g., using `python-dateutil`. My apologies!

Re. this plan:

> I'm going to keep the current behavior with python-dateutil and leave out the strptime fastpath, unless it is super important to someone. At some point we could cache (per-column) the internal parser that dateutil.parser discovers.

🔼 that sounds good to me!

And thanks for the heads-up re. the Threads Sheet. I hadn't explored that before.

Cheers,
Jeremy
